{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Examples of the BioSCRAPE package\n",
    "## Biocircuit Stochastic Simulation of Single Cell Reactions and Parameter Estimation\n",
    "The purpose of this Python notebook is twofold.\n",
    "\n",
    "1. The first is to serve as a quick start guide where you should be able to get started with the package by simply looking at the examples here and copying them to your liking.\n",
    "\n",
    "2. The second is as a unit testing replacement. It is hard to unit test stochastic algorithms as the output may not (and should not) be the same thing every time. Therefore, instead, if all the examples included below work well, then you can assume that the package installed correctly and is working fine.\n",
    "\n",
    "Before, getting started, we start by doing some basic plotting configuration and importing the numpy library. Advanced users can modify this to their liking."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "#%config InlineBackend.figure_f.ormats=['svg']\n",
    "color_list = ['r', 'k', 'b','g','y','m','c']\n",
    "mpl.rc('axes', prop_cycle=(mpl.cycler('color', color_list) ))\n",
    "mpl.rc('xtick', labelsize=12) \n",
    "mpl.rc('ytick', labelsize=12)\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Bioscrape Models\n",
    "## Chemical Reactions\n",
    "Bioscrape models consist of a set of species and a set of reactions (delays will be discussed later). These models can be simulated either stochastically via SSA or deterministically as an ODE. Each reaction is of the form\n",
    "\n",
    "$\\{INPUTS\\} \\xrightarrow[]{\\rho(.)} \\{OUTPUTS\\}$\n",
    "\n",
    "Here, INPUTS represent a multiset of input species and OUTPUTS represents a multiset of output species. The function $\\rho(.)$ is either a deterministic rate function or a stochastic propensity. Propensities are identified by their name and require parameter dictionary with the appropriate parameters. The following functions are supported:\n",
    "\n",
    "1. \"massaction: $\\rho(S) = k \\Pi_{s} s^{I_s}$. Required parameters: \"k\" the rate constant. Note: for stochastic simulations mass action propensities are $\\rho(S) = \\frac{1}{V} k \\Pi_{s} s!/(s - I_s)!$ where $V$ is the volume.\n",
    "2. \"positivehill\": $\\rho(s) = k \\frac{s^n}{(K^n+s^n)}$. Requried parameters: rate constant \"k\", offset \"K\", hill coefficient \"n\", hill species \"s\".\n",
    "3. \"negativehill\": $\\rho(s) = k \\frac{1}{(K^n+s^n)}$. Requried parameters: rate constant \"k\", offset \"K\", hill coefficient \"n\", hill species \"s\".\n",
    "4. \"proportionalpositivehill\": $\\rho(s) = k d \\frac{s^n}{(K^n+s^n)}$. Requried parameters: rate constant \"k\", offset \"K\", hill coefficient \"n\", hill species \"s\", propritional species \"d\".\n",
    "5. \"proportionalnegativehill\": $\\rho(s) = k d \\frac{1}{(K^n+s^n)}$. Requried parameters: rate constant \"k\", offset \"K\", hill coefficient \"n\", hill species \"s\", propritional species \"d\".\n",
    "6. \"general\": $\\rho(s) = f(s)$ where $f$ can be any algebraic function typed as a string. Required parameters: \"rate\" an algebraic expression including species and model parameters written as a string.\n",
    "\n",
    "More details on all these propensity types can be found in the <a href=\"https://github.com/ananswam/bioscrape/wiki/Propensities\">wiki documentation</a> \n",
    "\n",
    "## Transcription Translation Example\n",
    "First, the following model of transcription and translation will be created programatically. There are three chemical species: $G$ is a gene, $T$ is a transcript, $X$ is a protein.\n",
    "\n",
    "1. $G \\xrightarrow[]{\\rho_{tx}(G, I)} G+T$; $\\rho_{tx}(G, I) = G k_{tx}\\frac{I^{n}}{K_{I}^{n}+I^{n}}$, $I$ is an inducer.\n",
    "2. $T \\xrightarrow[]{\\rho_{tl}(T)} T+X$; $\\rho_{tl}(T) = k_{tl} \\frac{T}{K_{R} + T}$, $k_{tl}$ and $K_R$ model effects due to ribosome saturation.\n",
    "3. $T \\xrightarrow[]{\\delta} \\varnothing$; massaction kinetics at rate $\\delta$.\n",
    "4. $X \\xrightarrow[]{\\delta} \\varnothing$; massaction kinetics at rate $\\delta$.\n",
    "\n",
    "The first reaction uses a proportional positive hill function as its rate function to represent induction. The second reaction uses a positive hill function function to represent ribosome saturation. The third and fourth reactions reaction represent degredation via dilution. No delays will be included in this model. This model is constructed below and simulated both stochastically and deterministically."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\apand\\Anaconda3\\lib\\site-packages\\ipykernel_launcher.py:24: UserWarning: The following species are uninitialized and their value has been defaulted to 0: T, X, \n"
     ]
    }
   ],
   "source": [
    "from bioscrape.simulator import py_simulate_model\n",
    "from bioscrape.types import Model\n",
    "\n",
    "#Create a list of species names (strings)\n",
    "species = [\"G\", \"T\", \"X\", \"I\"]\n",
    "\n",
    "#create a list of parameters in the form (param_name[string], param_val[number])\n",
    "params = [(\"ktx\", 1.5), (\"ktl\", 10.0), (\"KI\", 10), (\"n\", 2.0), (\"KR\", 20), (\"delta\", .1)]\n",
    "\n",
    "#create reaction tuples in the form:\n",
    "#(Inputs[string list], Outputs[string list], propensity_type[string], propensity_dict {propensity_param:model_param})\n",
    "rxn1 = ([\"G\"], [\"G\", \"T\"], \"proportionalhillpositive\", {\"d\":\"G\", \"s1\":\"I\", \"k\":\"ktx\", \"K\":\"KI\", \"n\":\"n\"})\n",
    "rxn2 = ([\"T\"], [\"T\", \"X\"], \"hillpositive\", {\"s1\":\"T\", \"k\":\"ktl\", \"K\":\"KR\", \"n\":1}) #Notice that parameters can also take numerical values instead of being named directly\n",
    "rxn3 = ([\"T\"], [], \"massaction\", {\"k\":\"delta\"})\n",
    "rxn4 = ([\"X\"], [], \"massaction\", {\"k\":\"delta\"})\n",
    "\n",
    "#Create a list of all reactions\n",
    "rxns = [rxn1, rxn2, rxn3, rxn4]\n",
    "\n",
    "#create an initial condition dictionary species not included in the dictionary will default to 0\n",
    "x0 = {\"G\":1, \"I\":10}\n",
    "\n",
    "#Instaniate the Model object\n",
    "M = Model(species = species, parameters = params, reactions = rxns, initial_condition_dict = x0)\n",
    "\n",
    "#Simulate the Model deterministically\n",
    "timepoints = np.arange(0, 150, .1)\n",
    "results_det = py_simulate_model(timepoints, Model = M) #Returns a Pandas DataFrame\n",
    "\n",
    "#Simulate the Model Stochastically\n",
    "results_stoch = py_simulate_model(timepoints, Model = M, stochastic = True)\n",
    "\n",
    "#Plot the results\n",
    "plt.figure(figsize = (12, 4))\n",
    "plt.subplot(121)\n",
    "plt.title(\"Transcript T\")\n",
    "plt.plot(timepoints, results_det[\"T\"], label = \"deterministic\")\n",
    "plt.plot(timepoints, results_stoch[\"T\"], label = \"stochastic\")\n",
    "plt.legend()\n",
    "plt.xlabel(\"Time\")\n",
    "\n",
    "plt.subplot(122)\n",
    "plt.title(\"Protein X\")\n",
    "plt.plot(timepoints, results_det[\"X\"], label = \"deterministic\")\n",
    "plt.plot(timepoints, results_stoch[\"X\"], label = \"stochastic\")\n",
    "plt.legend()\n",
    "plt.xlabel(\"Time\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Adding Delays\n",
    "In stochastic simulations, bioscrape also supports delay. In a delay reaction, delay inputs/outputs are consumed/produced after some amount of delay. Reactions may have a mix of delay and non-delay inputs and outputs. Bioscrape innately supports a number of delay-types:\n",
    "\n",
    "1. fixed: constant delay with parameter \"delay\".\n",
    "2. Gaussian: gaussian distributed delay with parameters \"mean\" and \"std\".\n",
    "3. Gamma: gamma distributed delay with shape parameter \"k\" and scale parameter \"theta\".\n",
    "\n",
    "In the following example model, the following delays are added to the transcription and translation reactions described above and then simulated stochastically. Note that delays and delay inputs/outputs will be ignored if a model with delays is simulated deterministically."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from bioscrape.simulator import py_simulate_model\n",
    "from bioscrape.types import Model\n",
    "\n",
    "#create reaction tuples with delays require additional elements. They are of the form:\n",
    "#(Inputs[string list], Outputs[string list], propensity_type[string], propensity_dict {propensity_param:model_param},\n",
    "# delay_type[string], DelayInputs [string list], DelayOutputs [string list], delay_param_dictionary {delay_param:model_param}).\n",
    "\n",
    "rxn1d = ([\"G\"], [\"G\"], \"proportionalhillpositive\", {\"d\":\"G\", \"s1\":\"I\", \"k\":\"ktx\", \"K\":\"KI\", \"n\":\"n\"},\n",
    "       \"gaussian\", [], [\"T\"], {\"mean\":10.0, \"std\":1.0})\n",
    "rxn2d = ([\"T\"], [\"T\"], \"hillpositive\", {\"s1\":\"T\", \"k\":\"ktl\", \"K\":\"KR\", \"n\":1},\n",
    "        \"gamma\", [], [\"X\"], {\"k\":10.0, \"theta\":3.0})\n",
    "\n",
    "#Reactions 3 and 4 remain unchanged\n",
    "rxns_delay = [rxn1d, rxn2d, rxn3, rxn4]\n",
    "\n",
    "#Instaniate the Model object, species, params, and x0 remain unchanged from the previous example\n",
    "M_delay = Model(species = species, parameters = params, reactions = rxns_delay, initial_condition_dict = x0)\n",
    "\n",
    "#Simulate the Model with delay\n",
    "results_delay = py_simulate_model(timepoints, Model = M_delay, stochastic = True, delay = True)\n",
    "\n",
    "#Plot the results\n",
    "plt.figure(figsize = (12, 4))\n",
    "plt.subplot(121)\n",
    "plt.title(\"Transcript T\")\n",
    "plt.plot(timepoints, results_det[\"T\"], label = \"deterministic (no delay)\")\n",
    "plt.plot(timepoints, results_stoch[\"T\"], label = \"stochastic (no delay)\")\n",
    "plt.plot(timepoints, results_delay[\"T\"], label = \"stochastic (with delay)\")\n",
    "plt.legend()\n",
    "plt.xlabel(\"Time\")\n",
    "\n",
    "plt.subplot(122)\n",
    "plt.title(\"Protein X\")\n",
    "plt.plot(timepoints, results_det[\"X\"], label = \"deterministic (no delay)\")\n",
    "plt.plot(timepoints, results_stoch[\"X\"], label = \"stochastic (no delay)\")\n",
    "plt.plot(timepoints, results_delay[\"X\"], label = \"stochastic (with delay)\")\n",
    "\n",
    "plt.legend()\n",
    "plt.xlabel(\"Time\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Adding Rules\n",
    "In deterministic and stochastic simulations, bioscrape also supports rules which can be used to set species or parameter values during the simulation. Rules are updated every simulation timepoint - and therefore the model may be sensitive to how the timepoint spacing.\n",
    "\n",
    "The following example two rules will be added to the above model (without delay).\n",
    "\n",
    "1. $I = I_0 H(T)$ where $H$ is the step function. Represents the addition of the inducer I at concentrion $I_0$ some time T. Prior to t=T, I is not present.\n",
    "2. $S = M \\frac{X}{1+aX}$ represents a saturating signal detected from the species X via some sort of sensor.\n",
    "\n",
    "Rules can also be used for quasi-steady-state or quasi-equilibrium approximations, computing parameters on during the simulation, and much more!\n",
    "\n",
    "There are two main types of rules:\n",
    "1. \"additive\": used for calculating the total of many species. Rule 'equation' must be in the form $s_0 = s_1 + s_2 ...$ where $s_i$ each represents a species string.\n",
    "2. \"assignment\": a general rule type of with 'equation' form $v = f(s, p)$ where $v$ can be either a species or parameter which is assigned the value $f(s, p)$ where $s$ are all the species and $p$ are all the parameters in the model and $f$ is written as an string."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create new parameters for rule 1. Model is now being modified\n",
    "M.create_parameter(\"I0\", 10) #Inducer concentration\n",
    "M.create_parameter(\"T\", 25) #Initial time inducer is added\n",
    "\n",
    "#Create rule 1:\n",
    "#NOTE Rules can also be passed into the Model constructor as a list of tuples [(\"rule_type\", {\"equation\":\"eq string\"})]\n",
    "M.create_rule(\"assignment\", {\"equation\":\"I = _I0*Heaviside(t-_T)\"}) #\"_\" must be placed before param names, but not species.\n",
    "\n",
    "#Rule 2 will use constants in equations instead of new parameters. \n",
    "M.create_rule(\"assignment\", {\"equation\":\"S = 50*X/(1+.2*X)\"}) #Species S is added automatically\n",
    "\n",
    "#reset the initial concentration of the inducer to 0\n",
    "M.set_species({\"I\":0})\n",
    "\n",
    "#Simulate the Model deterministically\n",
    "timepoints = np.arange(0, 150, 1.0)\n",
    "results_det = py_simulate_model(timepoints, Model = M) #Returns a Pandas DataFrame\n",
    "\n",
    "#Simulate the Model Stochastically\n",
    "results_stoch = py_simulate_model(timepoints, Model = M, stochastic = True)\n",
    "\n",
    "#Plot the results\n",
    "plt.figure(figsize = (12, 8))\n",
    "plt.subplot(223)\n",
    "plt.title(\"Transcript T\")\n",
    "plt.plot(timepoints, results_det[\"T\"], label = \"deterministic\")\n",
    "plt.plot(timepoints, results_stoch[\"T\"], label = \"stochastic\")\n",
    "plt.legend()\n",
    "\n",
    "plt.subplot(224)\n",
    "plt.title(\"Protein X\")\n",
    "plt.plot(timepoints, results_det[\"X\"], label = \"deterministic\")\n",
    "plt.plot(timepoints, results_stoch[\"X\"], label = \"stochastic\")\n",
    "plt.legend()\n",
    "\n",
    "plt.subplot(221)\n",
    "plt.title(\"Inducer I\")\n",
    "plt.plot(timepoints, results_det[\"I\"], label = \"deterministic\")\n",
    "plt.plot(timepoints, results_stoch[\"I\"], label = \"stochastic\")\n",
    "plt.legend()\n",
    "plt.xlabel(\"Time\")\n",
    "\n",
    "plt.subplot(222)\n",
    "plt.title(\"Signal S\")\n",
    "plt.plot(timepoints, results_det[\"S\"], label = \"deterministic\")\n",
    "plt.plot(timepoints, results_stoch[\"S\"], label = \"stochastic\")\n",
    "plt.legend()\n",
    "plt.xlabel(\"Time\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SBML Support : Saving and Loading Bioscrape Models via [SBML](http://sbml.org)\n",
    "Models can be saved and loaded as SBML. Here we will save and load the transcription translation model to a SBML file. Delays, compartments, function definitions, and other non-standard SBML is not supported. \n",
    "Once a model has been loaded, it can be accessed and modified via the API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "M.write_sbml_model('models/txtl_model_sbml.xml')\n",
    "f = open('models/txtl_model_sbml.xml')\n",
    "print(\"Bioscrape Model converted to SBML:\\n\", f.read())\n",
    "\n",
    "M_loaded = Model(sbml_filename = 'models/txtl_model_sbml.xml')\n",
    "\n",
    "#Change the induction time\n",
    "#NOTE That changing a model loaded from xml will not change the underlying XML.\n",
    "M_loaded.set_parameter(\"T\", 50)\n",
    "\n",
    "#Simulate the Model deterministically\n",
    "timepoints = np.arange(0, 150, 1.0)\n",
    "results_det = py_simulate_model(timepoints, Model = M_loaded) #Returns a Pandas DataFrame\n",
    "\n",
    "#Simulate the Model Stochastically\n",
    "results_stoch = py_simulate_model(timepoints, Model = M_loaded, stochastic = True)\n",
    "\n",
    "#Plot the results\n",
    "plt.figure(figsize = (12, 8))\n",
    "plt.subplot(223)\n",
    "plt.title(\"Transcript T\")\n",
    "plt.plot(timepoints, results_det[\"T\"], label = \"deterministic\")\n",
    "plt.plot(timepoints, results_stoch[\"T\"], label = \"stochastic\")\n",
    "plt.legend()\n",
    "\n",
    "plt.subplot(224)\n",
    "plt.title(\"Protein X\")\n",
    "plt.plot(timepoints, results_det[\"X\"], label = \"deterministic\")\n",
    "plt.plot(timepoints, results_stoch[\"X\"], label = \"stochastic\")\n",
    "plt.legend()\n",
    "\n",
    "plt.subplot(221)\n",
    "plt.title(\"Inducer I\")\n",
    "plt.plot(timepoints, results_det[\"I\"], label = \"deterministic\")\n",
    "plt.plot(timepoints, results_stoch[\"I\"], label = \"stochastic\")\n",
    "plt.legend()\n",
    "plt.xlabel(\"Time\")\n",
    "\n",
    "plt.subplot(222)\n",
    "plt.title(\"Signal S\")\n",
    "plt.plot(timepoints, results_det[\"S\"], label = \"deterministic\")\n",
    "plt.plot(timepoints, results_stoch[\"S\"], label = \"stochastic\")\n",
    "plt.legend()\n",
    "plt.xlabel(\"Time\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Saving and Loading Bioscrape Models via Bioscrape XML\n",
    "\n",
    "Models can be saved and loaded as Bioscrape XML. Here we will save and load the transcription translation model and display the bioscrape XML underneath. Once a model has been loaded, it can be accessed and modified via the API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "M.write_bioscrape_xml('models/txtl_model.xml')\n",
    "f = open('models/txtl_model.xml')\n",
    "print(\"Bioscrape Model XML:\\n\", f.read())\n",
    "\n",
    "M_loaded = Model('models/txtl_model.xml')\n",
    "\n",
    "#Change the induction time\n",
    "#NOTE That changing a model loaded from xml will not change the underlying XML.\n",
    "M_loaded.set_parameter(\"T\", 50)\n",
    "\n",
    "#Simulate the Model deterministically\n",
    "timepoints = np.arange(0, 150, 1.0)\n",
    "results_det = py_simulate_model(timepoints, Model = M_loaded) #Returns a Pandas DataFrame\n",
    "\n",
    "#Simulate the Model Stochastically\n",
    "results_stoch = py_simulate_model(timepoints, Model = M_loaded, stochastic = True)\n",
    "\n",
    "#Plot the results\n",
    "plt.figure(figsize = (12, 8))\n",
    "plt.subplot(223)\n",
    "plt.title(\"Transcript T\")\n",
    "plt.plot(timepoints, results_det[\"T\"], label = \"deterministic\")\n",
    "plt.plot(timepoints, results_stoch[\"T\"], label = \"stochastic\")\n",
    "plt.legend()\n",
    "\n",
    "plt.subplot(224)\n",
    "plt.title(\"Protein X\")\n",
    "plt.plot(timepoints, results_det[\"X\"], label = \"deterministic\")\n",
    "plt.plot(timepoints, results_stoch[\"X\"], label = \"stochastic\")\n",
    "plt.legend()\n",
    "\n",
    "plt.subplot(221)\n",
    "plt.title(\"Inducer I\")\n",
    "plt.plot(timepoints, results_det[\"I\"], label = \"deterministic\")\n",
    "plt.plot(timepoints, results_stoch[\"I\"], label = \"stochastic\")\n",
    "plt.legend()\n",
    "plt.xlabel(\"Time\")\n",
    "\n",
    "plt.subplot(222)\n",
    "plt.title(\"Signal S\")\n",
    "plt.plot(timepoints, results_det[\"S\"], label = \"deterministic\")\n",
    "plt.plot(timepoints, results_stoch[\"S\"], label = \"stochastic\")\n",
    "plt.legend()\n",
    "plt.xlabel(\"Time\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SBML Compatibility\n",
    "\n",
    "The next cell imports a model from an SBML file and then simulates it using a deterministic simulation. There are limitations to SBML compatibility.\n",
    "\n",
    "1. Cannot support delays or events when reading in SBML files. Events will be ignored and a warning will be printed out.\n",
    "2. SBML reaction rates must be in a format such that when the reaction rates are converted to a string formula, sympy must be able to parse the formula. This will work fine for usual PEMDAS rates. This will fail for complex function definitions and things like that.\n",
    "3. Species will be initialized to their initialAmount field when it is nonzero. If the initialAmount is zero, then the initialConcentration will be used instead.\n",
    "4. Multiple compartments or anything related to having compartments will not be supported. No warnings will be provided for this.\n",
    "5. Assignment rules are supported, but any other type of rule will be ignored and an associated warning will be printed out.\n",
    "6. Parameter names must start with a letter and be alphanumeric, same for species names. Furthermore, log, exp, abs, heaviside, and other associated keywords for functions are not allowed to be variable names. When in doubt, just pick something else :)\n",
    "\n",
    "Below, we first plot out the simulation results for an SBML model where a species X0 goes to a final species X1 through an enymatic process."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from bioscrape.sbmlutil import import_sbml\n",
    "M_sbml = import_sbml('models/sbml_test.xml')\n",
    "timepoints = np.linspace(0,100,1000)\n",
    "\n",
    "result = py_simulate_model(timepoints, Model = M_sbml)\n",
    "\n",
    "plt.figure()\n",
    "for s in M_sbml.get_species_list():\n",
    "    plt.plot(timepoints, result[s], label = s)\n",
    "\n",
    "plt.legend()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Deterministic and Stochastic Simulation of the Repressilator\n",
    "\n",
    "We  plot out the repressilator model found <a href=\"http://www.ebi.ac.uk/biomodels-main/BIOMD0000000012\">here</a>. This model generates oscillations as expected. Highlighting the utility of this package, we then with a single line of code switch to a stochastic simulation and note that the amplitudes of each burst become noisy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Repressilator deterministic example\n",
    "import bioscrape\n",
    "M_represillator = bioscrape.sbmlutil.import_sbml('models/repressilator_sbml.xml')\n",
    "\n",
    "#Simulate Deterministically and Stochastically\n",
    "timepoints = np.linspace(0,700,10000)\n",
    "result_det = py_simulate_model(timepoints, Model = M_represillator)\n",
    "result_stoch = py_simulate_model(timepoints, Model = M_represillator, stochastic = True)\n",
    "\n",
    "#Plot Results\n",
    "plt.figure(figsize = (12, 8))\n",
    "for i in range(len(M_represillator.get_species_list())):\n",
    "    s = M_represillator.get_species_list()[i]\n",
    "    plt.plot(timepoints, result_det[s], color = color_list[i], label = \"Deterministic \"+s)\n",
    "    plt.plot(timepoints, result_stoch[s], \":\", color = color_list[i], label = \"Stochastic \"+s)\n",
    "\n",
    "plt.title('Repressilator Model')\n",
    "plt.xlabel('Time')\n",
    "plt.ylabel('Amount')\n",
    "plt.legend()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Simulating Cell Lineages\n",
    "See Lineage Examples for details on how to simulate single cells and lineages of cells."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
